google/cloud/spanner_v1/merge_result_set.py

```diff
@@ -0,0 +1,133 @@
+        METADATA_LOCK.release()
+        merged_result_set.metadata_semaphore.release()
+
+
+class PartitionExecutor:
+    """
+    Executor that executes single partition on a separate thread and inserts
+    rows in the queue
+    """
+
+    def __init__(self, batch_snapshot, partition_id, merged_result_set):
+        self._batch_snapshot: BatchSnapshot = batch_snapshot
+        self._partition_id = partition_id
+        self._merged_result_set: MergedResultSet = merged_result_set
+        self._queue: Queue[PartitionExecutorResult] = merged_result_set._queue
+
+    def run(self):
+        try:
+            results = self._batch_snapshot.process_query_batch(self._partition_id)
+            merged_result_set = self._merged_result_set
+            for row in results:
+                if merged_result_set._metadata is None:
+                    _set_metadata(merged_result_set, results)
+                self._queue.put(PartitionExecutorResult(data=row))
+            # Special case: The result set did not return any rows.
+            # Push the metadata to the merged result set.
+            if merged_result_set._metadata is None:
+                _set_metadata(merged_result_set, results)
+        except Exception as ex:
+            self._queue.put(PartitionExecutorResult(exception=ex))
+        finally:
+            # Emit a special 'is_last' result to ensure that the MergedResultSet
+            # is not blocked on a queue that never receives any more results.
+            self._queue.put(PartitionExecutorResult(is_last=True))
+
+
+@dataclass
+class PartitionExecutorResult:
+    data: Any = None
+    exception: Exception = None
+    is_last: bool = False
+
+
+class MergedResultSet:
+    """
+    Executes multiple partitions on different threads and then combines the
+    results from multiple queries using a synchronized queue. The order of the
+    records in the MergedResultSet is not guaranteed.
+    """
+
+    def __init__(self, batch_snapshot, partition_ids, max_parallelism):
+        self._exception = None
+        self.metadata_semaphore = Semaphore(0)
+
+        partition_ids_count = len(partition_ids)
+        self._finished_count_down_latch = partition_ids_count
+        parallelism = min(MAX_PARALLELISM, partition_ids_count)
+        if max_parallelism != 0:
+            parallelism = min(partition_ids_count, max_parallelism)
+        self._queue = Queue(maxsize=QUEUE_SIZE_PER_WORKER * parallelism)
+
+        partition_executors = []
+        for partition_id in partition_ids:
+            partition_executors.append(
+                PartitionExecutor(batch_snapshot, partition_id, self)
+            )
+        executor = ThreadPoolExecutor(max_workers=parallelism)
+        for partition_executor in partition_executors:
+            executor.submit(partition_executor.run)
+        executor.shutdown(False)
+
+        self._metadata = None
+        # This will make sure that _metadata is set
+        self.metadata_semaphore.acquire()
```

Comment:
I think that this means that the `__init__` method will block here until the metadata has been set, right? I don't think we should do that. Instead, we should add `acquire()` and `release()` calls to the `def metadata(self)` property. That would cause calls to that property to block until the metadata are available. That means that there is no blocking if the application does not try to get the metadata. Now, it will always block for that.

Reply:
I was doing the same but it was not working and after spending quite some time I found that property `def metadata(self)` was called by cursor `description()` property 2 times and hence it was getting deadlocked so moved it to `__init__` method.

Changing it to `Event` (https://docs.python.org/3/library/threading.html#event-objects) and also moved the wait to `def metadata(self)` as its better suited here